File: res/jburkardt/people.sc.fsu.edu/~jburkardt/octave_src/stiff_ode/stiff_midpoint.m
function [ t, y ] = stiff_midpoint ( n )

%*****************************************************************************80
%
%% stiff_midpoint() uses the midpoint method on the stiff ODE.
%
%  Discussion:
%
%    y' = lambda * ( cos(t) - y )
%    y(t0) = y0
%
%  Licensing:
%
%    This code is distributed under the GNU LGPL license.
%
%  Modified:
%
%    06 April 2021
%
%  Author:
%
%    John Burkardt
%
%  Input:
%
%    integer N: the number of steps to take.
%
%  Output:
%
%    real T(N+1), Y(N+1): the times and estimated solutions.
%
  t = zeros ( n + 1, 1 );
  y = zeros ( n + 1, 1 );

  [ lambda, t0, y0 ] = stiff_parameters ( );

  tstop = 1.0;
  dt = ( tstop - t0 ) / n;

  t(1) = t0;
  y(1) = y0;

  for i = 1 : n
    t(i+1) = t(i) + dt;
    y(i+1) = ... 
      ( ...
        y(i) + lambda * 0.5 * dt * ( ...
                                     cos ( t(i) ) - y(i) + cos ( t(i+1) ) ...
                                   ) ...
      ) ...
      / ( 1.0 + lambda * 0.5 * dt );
  end

  return
end

File: res/jburkardt/people.sc.fsu.edu/~jburkardt/r_src/midpoint_test/stiff_midpoint.R
stiff_midpoint <- function ()

#*****************************************************************************80
#
## stiff_midpoint() uses midpoint() to solve the stiff ODE.
#
#  Licensing:
#
#    This code is distributed under the GNU LGPL license.
#
#  Modified:
#
#    28 April 2021
#
{
  cat ( "\n" )
  cat ( "stiff_midpoint\n" )
  cat ( " ", version$version.string, "\n" )
  cat ( "  Use midpoint() to solve the stiff ordinary differential equation.\n" )

  library ( "pracma" )

  source ( "/home/burkardt/public_html/r_src/midpoint/midpoint.R" ) 
  source ( "/home/burkardt/public_html/r_src/midpoint_test/stiff_deriv.R" )
  source ( "/home/burkardt/public_html/r_src/midpoint_test/stiff_exact.R" )  
  source ( "/home/burkardt/public_html/r_src/midpoint_test/stiff_parameters.R" )   

  parameters <- stiff_parameters ( )

  t0 <- parameters$t0
  y0 <- parameters$y0
  tstop <- parameters$tstop
  n <- 27

  sol <- midpoint ( stiff_deriv, t0, tstop, y0, n )
#
#  Time plot.
#
  t <- sol$t
  y <- sol$y

  ye <- stiff_exact ( t )

  png ( 'stiff_midpoint_plot.png' )
  plot ( t, y, type = "l", lwd = 2, col = "red",
       xlab = "", ylab = "", main = '"stiff midpoint: plot"')
  lines ( t, ye, type = "l", lwd = 2, col = "blue" )
  grid ( )
}
